### PR TITLE
Tabular: Added LightGBM DART model retrain support, added GPU support for CatBoost

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

*Description of changes:*

Added LightGBM DART model retrain support, added GPU support for CatBoost.

TODO: Benchmark

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-682-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-682/1/index.html

**Comment by szha** [comment]
Job PR-682-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-682/2/index.html

**Comment by jwmueller** [comment]
Later we should consider keeping `num_gpus = None` as default in `task.fit()`, and then allowing each model to autodetect GPU and utilize it when it sees `num_gpus=None`

**Comment by Innixma** [comment]
> Later we should consider keeping `num_gpus = None` as default in `task.fit()`, and then allowing each model to autodetect GPU and utilize it when it sees `num_gpus=None`

Yes, I had a similar idea. Either `num_gpus=None` or `num_gpus='auto'`. We need to very solid in terms of fallback options if the GPU doesn't work before we go this route though, so likely try/except checks for each model in case GPU training fails. This gets more tricky with models like LightGBM that need special installations to use GPU, but hopefully try/except will still work.

**Comment by Innixma** [comment]
What makes this even trickier is that GPU isn't simply a training speedup, it fundamentally changes how the models are trained and their resulting accuracy, often for the worse.

**Comment by Innixma** [comment]
Benchmark results show that while DART does help slightly when stack ensembling is disabled, it adds very little if anything to the stack ensemble results (weakest average results of the four LGBM variants). Since it is an expensive model to train, I have removed it from the default model set.

**Comment by szha** [comment]
Job PR-682-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-682/3/index.html

**Comment by jwmueller** [comment]
> What makes this even trickier is that GPU isn't simply a training speedup, it fundamentally changes how the models are trained and their resulting accuracy, often for the worse.

Right but that's why each model could choose whether or not to utilize the GPU based on auto-detection of GPU availability + possibly some dataset characteristics.  That flexibility seems a much better default.

File: autogluon/utils/tabular/ml/models/catboost/catboost_model.py [hunk 1]
```diff
@@ -146,6 +146,11 @@ def _fit(self, X_train, y_train, X_val=None, y_val=None, time_limit=None, **kwar
 
         params = self.params.copy()
         num_features = len(self.features)
+        if num_gpus != 0:
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/models/catboost/catboost_model.py:149]: add comment that only 1 GPU can be utilized?

File: autogluon/utils/tabular/ml/models/catboost/catboost_model.py [hunk 2]
```diff
@@ -146,6 +146,11 @@ def _fit(self, X_train, y_train, X_val=None, y_val=None, time_limit=None, **kwar
...
+            if 'task_type' not in params:
+                params['task_type'] = 'GPU'
+                # TODO: Confirm if GPU is used in HPO (Probably not)
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/models/catboost/catboost_model.py:152]: any idea why GPU isn't utilized during HPO?

File: autogluon/utils/tabular/ml/models/lgb/lgb_model.py [hunk 3]
```diff
@@ -74,6 +74,15 @@ def _fit(self, X_train=None, y_train=None, X_val=None, y_val=None, dataset_train
         gc.collect()
 
         num_boost_round = params.pop('num_boost_round', 1000)
+        dart_retrain = params.pop('dart_retrain', False)  # Whether to retrain the model to get optimal iteration if model is trained in 'dart' mode.
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/models/lgb/lgb_model.py:77]: For both of these hyperparams, I don't think we should have specify the default value here?  Shouldn't we just ensure all default values come from `hyperparameters/parameters.py`, so that they're very clearly documented?

**Review by jwmueller — APPROVED** [review]
Looks good, just have minor comments

File: autogluon/utils/tabular/ml/models/catboost/catboost_model.py (hunk 1 above)
Comment (Innixma) [reply autogluon/utils/tabular/ml/models/catboost/catboost_model.py:149]: I believe CatBoost can use multi-gpu.

File: autogluon/utils/tabular/ml/models/catboost/catboost_model.py (hunk 2 above)
Comment (Innixma) [reply autogluon/utils/tabular/ml/models/catboost/catboost_model.py:152]: Disjoint code calls, I don't think the num_gpus param is passed with HPO. In general, I'm in favor of moving num_gpus into params instead of fit, that gives us a lot more flexibility in terms of per-model GPU usage, and keeps model .fit args clean. It could be part of AG_args_fit.

File: autogluon/utils/tabular/ml/models/lgb/lgb_model.py (hunk 3 above)
Comment (Innixma) [reply autogluon/utils/tabular/ml/models/lgb/lgb_model.py:77]: dart_retrain is very niche, and probably will only be for internal usage. I agree in general however, and I think we can maybe clean that up. Do you suggest that if num_boost_round doesn't exist at this point in the code to throw an exception?

Comment (jwmueller) [reply autogluon/utils/tabular/ml/models/lgb/lgb_model.py:77]: Yep could just do `params['num_boost_round']` instead of `.pop()`
